README.md
# VESC firmware

[![License: GPL v3](https://img.shields.io/badge/License-GPLv3-blue.svg)](https://www.gnu.org/licenses/gpl-3.0)
[![Travis CI Status](https://travis-ci.com/vedderb/bldc.svg?branch=master)](https://travis-ci.com/vedderb/bldc)
[![Codacy Badge](https://api.codacy.com/project/badge/Grade/75e90ffbd46841a3a7be2a9f7a94c242)](https://www.codacy.com/app/vedderb/bldc?utm_source=github.com&amp;utm_medium=referral&amp;utm_content=vedderb/bldc&amp;utm_campaign=Badge_Grade)
[![Contributors](https://img.shields.io/github/contributors/vedderb/bldc.svg)](https://github.com/vedderb/bldc/graphs/contributors)
[![Watchers](https://img.shields.io/github/watchers/vedderb/bldc.svg)](https://github.com/vedderb/bldc/watchers)
[![Stars](https://img.shields.io/github/stars/vedderb/bldc.svg)](https://github.com/vedderb/bldc/stargazers)
[![Forks](https://img.shields.io/github/forks/vedderb/bldc.svg)](https://github.com/vedderb/bldc/network/members)

An open source motor controller firmware.

This is the source code for the VESC DC/BLDC/FOC controller. Read more at
[https://vesc-project.com/](https://vesc-project.com/)

## Supported boards

All of them!

Check the supported boards by typing `make`

```
[Firmware]
     fw   - Build firmware for default target
                            supported boards are: 100_250 100_250_no_limits 100_500...
```

There are also many other options that can be changed in [conf_general.h](conf_general.h).

## Prerequisites

### On Ubuntu (Linux)/macOS
- Tools: `git`, `wget`, and `make`
- Additional Linux requirements: `libgl-dev` and `libxcb-xinerama0`
- Helpful Ubuntu commands:
```bash
sudo apt install git build-essential libgl-dev libxcb-xinerama0 wget git-gui
```
- Helpful macOS tools: 

```bash
brew install stlink
brew install openocd
```

### On Windows
- Chocolately: https://chocolatey.org/install
- Git: https://git-scm.com/download/win. Make sure to click any boxes to add Git to your Environment (aka PATH)

## Install Dev environment and build

### On Ubuntu (Linux)/MacOS
Open up a terminal
1.  `git clone http://github.com/vedderb/bldc.git`
2.  `cd bldc`
3.  Continue with [On all platforms](#on-all-platforms)

### On Windows

1.  Open up a Windows Powershell terminal (Resist the urge to run Powershell as administrator, that will break things)
2.  Type `choco install make`
3.  `git clone http://github.com/vedderb/bldc`
4.  `cd bldc`
5.  Continue with [On all platforms](#on-all-platforms)

### On all platforms

1.  `git checkout origin/master`
2.  `make arm_sdk_install`
3.  `make` <-- Pick out the name of your target device from the supported boards list. For instance, I have a Trampa **VESC 100/250**, so my target is `100_250`
4.   `make 100_250` <-- This will build the **VESC 100/250** firmware and place it into the `bldc/builds/100_250/` directory

## Other tools

**Linux Optional - Add udev rules to use the stlink v2 programmer without being root**
```bash
wget vedder.se/Temp/49-stlinkv2.rules
sudo mv 49-stlinkv2.rules /etc/udev/rules.d/
sudo udevadm trigger
```

## IDE
### Prerequisites
#### On macOS/Linux

- `python3`, and `pip`

#### On Windows
- Python 3: https://www.python.org/downloads/. Make sure to click the box to add Python3 to your Environment.

### All platforms

1.  `pip install aqtinstall`
2.  `make qt_install`
3.  Open Qt Creator IDE installed in `tools/Qt/Tools/QtCreator/bin/qtcreator`
4.  With Qt Creator, open the vesc firmware Qt Creator project, named vesc.pro. You will find it in `Project/Qt Creator/vesc.pro`
5.  The IDE is configured by default to build 100_250 firmware, this can be changed in the bottom of the left panel, there you will find all hardware variants supported by VESC

## Upload to VESC
### Method 1 - Flash it using an STLink SWD debugger

1.  Build and flash the [bootloader](https://github.com/vedderb/bldc-bootloader) first
2.  Then `_flash` to the target of your choice. So for instance, for the VESC 100/250: 
```bash
make 100_250_flash
```

### Method 2 - Upload Firmware via VESC tool through USB

1.  Clone and build the firmware in **.bin** format as in the above Build instructions

In VESC tool

2.  Connect to the VESC
3.  Navigate to the Firmware tab on the left side menu 
4.  Click on Custom file tab
5.  Click on the folder icon to select the built firmware in .bin format (e.g. `build/100_250/100_250.bin`)

##### [ Reminder : It is normal to see VESC disconnects during the firmware upload process ]  
#####  **[ Warning : DO NOT DISCONNECT POWER/USB to VESC during the upload process, or you will risk bricking your VESC ]**  
#####  **[ Warning : ONLY DISCONNECT your VESC 10s after the upload loading bar completed and "FW Upload DONE" ]**

6.  Press the upload firmware button (downward arrow) on the bottom right to start upload the selected firmware.
7.  Wait for **10s** after the loading bar completed (Warning: unplug sooner will risk bricking your VESC)
8.  The VESC will disconnect itself after new firmware is uploaded.

## In case you bricked your VESC
you will need to upload a new working firmware to the VESC.  
However, to upload a firmware to a bricked VESC, you have to use a SWD Debugger.


## Contribute

Head to the [forums](https://vesc-project.com/forum) to get involved and improve this project.
Join the [Discord](https://discord.gg/JgvV5NwYts) for real-time support and chat

## Tags

Every firmware release has a tag. They are created as follows:

```bash
git tag -a [version] [commit] -m "VESC Firmware Version [version]"
git push --tags
```

## License

The software is released under the GNU General Public License version 3.0

# Luna X2 settings (by MoreDH)

These are the settings I use for trail riding.

## Bike CFG
```
Current Max					60.00 A
Cells						13s
Overvoltage					60.00 V
Undervoltage Cutoff Start	40.00 V
Undervoltage Cutoff End		39.00 V
Wheel Size					27.5"
Offset Correction			74.8 deg
Invert Motor Direction		Off
```

## Tune

```
Throttle Amps				100.0 A
PAS Amps					100.0 A
Power						1250 W
Throttle Response			0.5 s
Throttle Linearity			0.0 %
PAS Response				0.4 s
Field Weakening				0.0 A
Fixed Throttle Amps			Off
```

CHANGELOG
=== FW 6.00 (2022.03.04 MoreDH) ===
* Battery SOC% is now mapped from 4.19 to 3.0V.
* PAS pedal timeout 5 -> 3s.
* More consistent PAS loop times.
* PAS torque filtered new butterworth 2nd order low pass filter.
* PAS0-9 is now 3 zone: LOW-PAS012 (0.025, 0.05, 0.1) / NORMAL-PAS3456(0.2, 0.25, 0.3, 0.35) / POWER-PAS789(0.5, 0.7, 1).
* PAS pedal torque to motor output is progressive (-3.0/poly).
* Defaults: 30M shutdown, max poly throttle linearity, PAS ramp time 0.5/0.3s, no field weakening.

=== FW 6.00 ===
* Added stack checks.
* Release motor fix when cc_min_current is 0.
* Added support for NTC and PTC temperature sensors with custom resistance and base temperature.
* Added sandboxed lisp scripting using https://github.com/svenssonjoel/lispBM.
* Encoder driver split and rewrite to make it easier to support more encoders.
* Flash writes now work on lower MCU voltages too.
* Added fwinfo terminal command.
* CAN status messages are now configured using bitfields.
* There are two rates at which CAN status messages can be sent now.
* New silent HFI mode with much better performance at high load.
* Added app ADC min and max voltage limits for throttle fault detection.
* Another new silent HFI mode that is less sensitive to getting the inductance correct.
* Added openloop current boost parameter.
* Added openloop current max parameter.
* Added support for dynamic loading of C libraries.
* Added more observers:
 - mxlemming
 - ortega with flux linkage tracking
 - mxlemming with flux linkage tracking
* AS5x47 encoder support: https://github.com/vedderb/bldc/pull/511
* Disable BMS limit options.
* Added PT1000 temperature sensor support.
* APP ADC button bitfield and CC disable support.
* BissC encoder support: https://github.com/vedderb/bldc/pull/536
* Better detection failt handling and reporting: https://github.com/vedderb/bldc/pull/533
* TLE5012-support: https://github.com/vedderb/bldc/pull/551

=== FW 5.03 ===
* Fixed inductance measurement bug.
* Speed tracker windup protection.
* Phase filter support.
* Phase voltage offset calibration.
* Better current offset calibration.
* Added power switch commands.
* Synchronize observer state when running in open loop.
* Force oberver state magnitude above 50% of flux linkage. This prevents the motor from getting stuck and 'screaming'.
* Observer global convergence update. Helps tracking the motor through 0 speed.
* Added HFI start sensor mode.
* Added TEMP_SENSOR_KTY84_130.
* Major UAVCAN update. See: https://github.com/vedderb/bldc/pull/269
* Avoid numerical instability when mapping is done over a narrow range. See: https://github.com/vedderb/bldc/issues/262
* App Balance updates.
* Added servo_out_enable appconf option, so that the PPM port can be used to control servos with the default firmware.
* Better current controller windup protection.
* Field weakening support (experimental, be careful and use at your own risk).
* Use filtered current for axis decoupling.
* Odometer is now persistent after firmware updates.
* Added hw runtime counter (see hw_status in terminal)
* Added custom UI support.
* Limit hall sensor angle rate of change based on ERPM.
* Added p_pid_gain_dec_angle parameter.
* Low pass filter input voltage.
* Dual hardware CAN-scan fix.
* Dynamic QML-script write support.
* Use fast speed tracker for current controller.
* Disable motor for 5 seconds after flash operations.
* Added kill switch support.
* Added process derivative term to position controller.
* Added position PID-controller angle offset.
* Configurable PID controller rate.
* Added several AUX port modes.
* Added configurable safe start modes.
* Added fusion IMU filter.
* Added constant torque PAS mode.
* Correct scaling for resistance and inductance.
* Fixed inductance measurement bug with f_sw > 30k.
* Corrected inductance measurement algorithm.
* Fixed max power loss calculation.
* Better input current estimation.
* Added raw sampled data mode.
* Compensate inductance for motor saliency in observer.
* Added MTPA mode based on measured current.
* Faster overvoltage protection.
* Added statistics counters.
* Added configurable observer offset.

=== FW 5.02 ===
* IMU calibration improvement.
* Added COMM_GET_MCCONF_TEMP command.
* Added bidirectional current command to VESC remote.
* Fixed motor temperature reading on hw with ADC mux.
* Added speed PID input ramping option.
* Added LSM6DS3 IMU support.
* Added MTPA support. See See: https://github.com/vedderb/bldc/pull/179
* Added HW_HD75 support.
* NRF52 UICR write fix.
* App PPM rework. See https://github.com/vedderb/bldc/pull/192
* Added bm_reset terminal command.
* Added bm support for STM32F30x and STM32L47x.
* App Balance updates. See https://github.com/vedderb/bldc/pull/193
* Motor current now based on magnitude of both axes.
* Initial VESC BMS support.
* Hall sensor interpolation improvement.
* Made hall sensor filter configurable.
* Added locking time and ramp up time parameters to sensorless startup.
* Removed D axis current injection.
* Initial VESC IO-board support.
* Added hall sensor interpolation ERPM config option.
* Use fast speed estimator for RPM limit.
* Avoid accumulated rounding error when using PID position angle division.
* Added UAVCAN raw throttle drive mode (current or duty cycle control).
* Added MT6816 encoder support. See https://github.com/vedderb/bldc/pull/238
* Added modulation-based D axis current controller gain scaling. Addresses https://github.com/vedderb/bldc/pull/220
* Added PAS app. See: https://github.com/vedderb/bldc/pull/243
* Added 100k CAN-baudrate.
* Added 100k NTC temperature sensor support.

=== FW 5.01 ===
* Fixed PPM bug in previous release.

=== FW 5.00 ===
* Dual motor support. VESC-based controllers such as the focbox unity will now work.
* Fixed bug in cross BEMF decoupling.
* Disable CC decoupling during flux linkage measurement.
* Balance app updates. See: https://github.com/vedderb/bldc/pull/141
* Observer gain calculation update.

hwconf/luna/m600/mcconf_luna_m600.h
// This file is autogenerated by VESC Tool

#ifndef MCCONF_LUNA_M600_H_
#define MCCONF_LUNA_M600_H_

// PWM Mode
#define MCCONF_PWM_MODE 1

// Commutation Mode
#define MCCONF_COMM_MODE 0

// Motor Type
#define MCCONF_DEFAULT_MOTOR_TYPE 2

// Sensor Mode
#define MCCONF_SENSOR_MODE 0

// Motor Current Max
#define MCCONF_L_CURRENT_MAX 100

// Motor Current Max Brake
#define MCCONF_L_CURRENT_MIN -20

// Battery Current Max
#define MCCONF_L_IN_CURRENT_MAX 60

// Battery Current Max Regen
#define MCCONF_L_IN_CURRENT_MIN -20

// Absolute Maximum Current
#define MCCONF_L_MAX_ABS_CURRENT 200

// Max ERPM Reverse
#define MCCONF_L_RPM_MIN -100000

// Max ERPM
#define MCCONF_L_RPM_MAX 100000

// ERPM Limit Start
#define MCCONF_L_RPM_START 0.8

// Max ERPM Full Brake
#define MCCONF_L_CURR_MAX_RPM_FBRAKE 300

// Max ERPM Full Brake Current Control
#define MCCONF_L_CURR_MAX_RPM_FBRAKE_CC 1500

// Minimum Input Voltage
#define MCCONF_L_MIN_VOLTAGE 30

// Maximum Input Voltage
#define MCCONF_L_MAX_VOLTAGE 60

// Battery Voltage Cutoff Start
#define MCCONF_L_BATTERY_CUT_START 40

// Battery Voltage Cutoff End
#define MCCONF_L_BATTERY_CUT_END 39

// Slow ABS Current Limit
#define MCCONF_L_SLOW_ABS_OVERCURRENT 0

// MOSFET Temp Cutoff Start
#define MCCONF_L_LIM_TEMP_FET_START 85

// MOSFET Temp Cutoff End
#define MCCONF_L_LIM_TEMP_FET_END 95

// Motor Temp Cutoff Start
#define MCCONF_L_LIM_TEMP_MOTOR_START 85

// Motor Temp Cutoff End
#define MCCONF_L_LIM_TEMP_MOTOR_END 120

// Acceleration Temperature Decrease
#define MCCONF_L_LIM_TEMP_ACCEL_DEC 0

// Minimum Duty Cycle
#define MCCONF_L_MIN_DUTY 0.005

// Maximum Duty Cycle
#define MCCONF_L_MAX_DUTY 0.95

// Maximum Wattage
#define MCCONF_L_WATT_MAX 2500

// Maximum Braking Wattage
#define MCCONF_L_WATT_MIN -2500

// Max Current Scale
#define MCCONF_L_CURRENT_MAX_SCALE 1

// Min Current Scale
#define MCCONF_L_CURRENT_MIN_SCALE 1

// Duty Cycle Current Limit Start
#define MCCONF_L_DUTY_START 1

// Minimum ERPM
#define MCCONF_SL_MIN_RPM 150

// Minimum ERPM Integrator
#define MCCONF_SL_MIN_ERPM_CYCLE_INT_LIMIT 1100

// Max Brake Current at Direction Change
#define MCCONF_SL_MAX_FB_CURR_DIR_CHANGE 10

// Cycle Integrator Limit
#define MCCONF_SL_CYCLE_INT_LIMIT 62

// Phase Advance at BR ERPM
#define MCCONF_SL_PHASE_ADVANCE_AT_BR 0.8

// BR ERPM
#define MCCONF_SL_CYCLE_INT_BR 80000

// BEMF Coupling
#define MCCONF_SL_BEMF_COUPLING_K 600

// Hall Table [0]
#define MCCONF_HALL_TAB_0 -1

// Hall Table [1]
#define MCCONF_HALL_TAB_1 1

// Hall Table [2]
#define MCCONF_HALL_TAB_2 3

// Hall Table [3]
#define MCCONF_HALL_TAB_3 2

// Hall Table [4]
#define MCCONF_HALL_TAB_4 5

// Hall Table [5]
#define MCCONF_HALL_TAB_5 6

// Hall Table [6]
#define MCCONF_HALL_TAB_6 4

// Hall Table [7]
#define MCCONF_HALL_TAB_7 -1

// Sensorless ERPM Hybrid
#define MCCONF_HALL_ERPM 2000

// Current KP
#define MCCONF_FOC_CURRENT_KP 0.256

// Current KI
#define MCCONF_FOC_CURRENT_KI 57.59

// Zero Vector Frequency
#define MCCONF_FOC_F_ZV 30000

// Dead Time Compensation
#define MCCONF_FOC_DT_US 0.23

// Encoder Inverted
#define MCCONF_FOC_ENCODER_INVERTED 1

// Encoder Offset
#define MCCONF_FOC_ENCODER_OFFSET 400.0

// Encoder Ratio
#define MCCONF_FOC_ENCODER_RATIO 4

// Sensor Mode
#define MCCONF_FOC_SENSOR_MODE 1

// Speed Tracker Kp
#define MCCONF_FOC_PLL_KP 2000

// Speed Tracker Ki
#define MCCONF_FOC_PLL_KI 30000

// Motor Inductance (L)
#define MCCONF_FOC_MOTOR_L 0.00025605

// Motor Inductance Difference (Ld - Lq)
#define MCCONF_FOC_MOTOR_LD_LQ_DIFF 0.00011882

// Motor Resistance (R)
#define MCCONF_FOC_MOTOR_R 0.0576

// Motor Flux Linkage (λ)
#define MCCONF_FOC_MOTOR_FLUX_LINKAGE 0.013535

// Observer Gain (x1M)
#define MCCONF_FOC_OBSERVER_GAIN 5.46e+06

// Observer Gain At Minimum Duty
#define MCCONF_FOC_OBSERVER_GAIN_SLOW 0.05

// Observer Offset
#define MCCONF_FOC_OBSERVER_OFFSET -1

// Duty Downramp Kp
#define MCCONF_FOC_DUTY_DOWNRAMP_KP 10

// Duty Downramp Ki
#define MCCONF_FOC_DUTY_DOWNRAMP_KI 200

// Start Current Decrease
#define MCCONF_FOC_START_CURR_DEC 1

// Start Current Decrease ERPM
#define MCCONF_FOC_START_CURR_DEC_RPM 2500

// Openloop ERPM
#define MCCONF_FOC_OPENLOOP_RPM 1500

// Openloop ERPM at Min Current
#define MCCONF_FOC_OPENLOOP_RPM_LOW 0

// D Axis Gain Scaling Start
#define MCCONF_FOC_D_GAIN_SCALE_START 1

// D Axis Gain Scaling at Max Mod
#define MCCONF_FOC_D_GAIN_SCALE_MAX_MOD 0.2

// Openloop Hysteresis
#define MCCONF_FOC_SL_OPENLOOP_HYST 0.1

// Openloop Lock Time
#define MCCONF_FOC_SL_OPENLOOP_T_LOCK 0

// Openloop Ramp Time
#define MCCONF_FOC_SL_OPENLOOP_T_RAMP 0.1

// Openloop Time
#define MCCONF_FOC_SL_OPENLOOP_TIME 0.05

// Openloop Current Boost
#define MCCONF_FOC_SL_OPENLOOP_BOOST_Q 0

// Openloop Current Max
#define MCCONF_FOC_SL_OPENLOOP_MAX_Q -1

// Hall Table [0]
#define MCCONF_FOC_HALL_TAB_0 255

// Hall Table [1]
#define MCCONF_FOC_HALL_TAB_1 255

// Hall Table [2]
#define MCCONF_FOC_HALL_TAB_2 255

// Hall Table [3]
#define MCCONF_FOC_HALL_TAB_3 255

// Hall Table [4]
#define MCCONF_FOC_HALL_TAB_4 255

// Hall Table [5]
#define MCCONF_FOC_HALL_TAB_5 255

// Hall Table [6]
#define MCCONF_FOC_HALL_TAB_6 255

// Hall Table [7]
#define MCCONF_FOC_HALL_TAB_7 255

// Hall Interpolation ERPM
#define MCCONF_FOC_HALL_INTERP_ERPM 500

// Sensorless ERPM
#define MCCONF_FOC_SL_ERPM 30000

// Sample in V0 and V7
#define MCCONF_FOC_SAMPLE_V0_V7 0

// High Current Sampling Mode
#define MCCONF_FOC_SAMPLE_HIGH_CURRENT 0

// Saturation Compensation Mode
#define MCCONF_FOC_SAT_COMP_MODE 2

// Saturation Compensation Factor
#define MCCONF_FOC_SAT_COMP 0

// Temp Comp
#define MCCONF_FOC_TEMP_COMP 0

// Temp Comp Base Temp
#define MCCONF_FOC_TEMP_COMP_BASE_TEMP 25

// Current Filter Constant
#define MCCONF_FOC_CURRENT_FILTER_CONST 0.1

// Current Controller Decoupling
#define MCCONF_FOC_CC_DECOUPLING 0

// Observer Type
#define MCCONF_FOC_OBSERVER_TYPE 3

// HFI Start Voltage
#define MCCONF_FOC_HFI_VOLTAGE_START 20

// HFI Run Voltage
#define MCCONF_FOC_HFI_VOLTAGE_RUN 4

// HFI Max Voltage
#define MCCONF_FOC_HFI_VOLTAGE_MAX 10

// HFI Gain
#define MCCONF_FOC_HFI_GAIN 0.3

// HFI Current Hysteresis
#define MCCONF_FOC_HFI_HYST 0

// Sensorless ERPM HFI
#define MCCONF_FOC_SL_ERPM_HFI 2000

// HFI Start Samples
#define MCCONF_FOC_HFI_START_SAMPLES 65

// HFI Observer Override Time
#define MCCONF_FOC_HFI_OBS_OVR_SEC 0.001

// HFI Samples
#define MCCONF_FOC_HFI_SAMPLES 1

// Run calibration at boot
#define MCCONF_FOC_OFFSETS_CAL_ON_BOOT 1

// Current Offset 0
#define MCCONF_FOC_OFFSETS_CURRENT_0 2048.0

// Current Offset 1
#define MCCONF_FOC_OFFSETS_CURRENT_1 2048.0

// Current Offset 2
#define MCCONF_FOC_OFFSETS_CURRENT_2 2048.0

// Voltage Offset 0
#define MCCONF_FOC_OFFSETS_VOLTAGE_0 0.0

// Voltage Offset 1
#define MCCONF_FOC_OFFSETS_VOLTAGE_1 0.0

// Voltage Offset 2
#define MCCONF_FOC_OFFSETS_VOLTAGE_2 0.0

// Voltage Offset Undriven 0
#define MCCONF_FOC_OFFSETS_VOLTAGE_UNDRIVEN_0 0

// Voltage Offset Undriven 1
#define MCCONF_FOC_OFFSETS_VOLTAGE_UNDRIVEN_1 0

// Voltage Offset Undriven 2
#define MCCONF_FOC_OFFSETS_VOLTAGE_UNDRIVEN_2 0

// Enable Phase Filters
#define MCCONF_FOC_PHASE_FILTER_ENABLE 0

// Disable Phase Filter Fault Code
#define MCCONF_FOC_PHASE_FILTER_DISABLE_FAULT 1

// Maximum ERPM for phase filters
#define MCCONF_FOC_PHASE_FILTER_MAX_ERPM 2500

// MTPA Algorithm Mode
#define MCCONF_FOC_MTPA_MODE 0

// Field Weakening Current Max
#define MCCONF_FOC_FW_CURRENT_MAX 0

// Field Weakening Duty Start
#define MCCONF_FOC_FW_DUTY_START 0.9

// Field Weakening Ramp Time
#define MCCONF_FOC_FW_RAMP_TIME 0.2

// Q Axis Current Factor
#define MCCONF_FOC_FW_Q_CURRENT_FACTOR 0.02

// Speed Tracker Position Source
#define MCCONF_FOC_SPEED_SOURCE 1

// Buffer Notification Length
#define MCCONF_GPD_BUFFER_NOTIFY_LEFT 200

// Buffer Sampling Interpolation
#define MCCONF_GPD_BUFFER_INTERPOL 0

// Current Filter Constant
#define MCCONF_GPD_CURRENT_FILTER_CONST 0.1

// Current KP
#define MCCONF_GPD_CURRENT_KP 0.03

// Current KI
#define MCCONF_GPD_CURRENT_KI 50

// PID Loop Rate
#define MCCONF_SP_PID_LOOP_RATE 5

// Speed PID Kp
#define MCCONF_S_PID_KP 0.004

// Speed PID Ki
#define MCCONF_S_PID_KI 0.004

// Speed PID Kd
#define MCCONF_S_PID_KD 0.0001

// Speed PID Kd Filter
#define MCCONF_S_PID_KD_FILTER 0.2

// Minimum ERPM
#define MCCONF_S_PID_MIN_RPM 100

// Allow Braking
#define MCCONF_S_PID_ALLOW_BRAKING 1

// Ramp eRPMs per second
#define MCCONF_S_PID_RAMP_ERPMS_S 1500

// Position PID Kp
#define MCCONF_P_PID_KP 0.03

// Position PID Ki
#define MCCONF_P_PID_KI 0

// Position PID Kd
#define MCCONF_P_PID_KD 0.0004

// Position PID Kd Process
#define MCCONF_P_PID_KD_PROC 0.00035

// Position PID Kd Filter
#define MCCONF_P_PID_KD_FILTER 0.2

// Position Angle Division
#define MCCONF_P_PID_ANG_DIV 1

// Gain Decrease Angle
#define MCCONF_P_PID_GAIN_DEC_ANGLE 0

// Position PID Offset Angle
#define MCCONF_P_PID_OFFSET 0

// Startup boost
#define MCCONF_CC_STARTUP_BOOST_DUTY 0.01

// Minimum Current
#define MCCONF_CC_MIN_CURRENT 0.05

// Current Controller Gain
#define MCCONF_CC_GAIN 0.0046

// Current Control Ramp Step Max
#define MCCONF_CC_RAMP_STEP 0.04

// Fault Stop Time
#define MCCONF_M_FAULT_STOP_TIME 500

// Duty Ramp Step Max
#define MCCONF_M_RAMP_STEP 0.02

// Current Backoff Gain
#define MCCONF_M_CURRENT_BACKOFF_GAIN 0.5

// Encoder counts
#define MCCONF_M_ENCODER_COUNTS 8192

// Sine Amplitude
#define MCCONF_M_ENCODER_SIN_AMP 1

// Cosine Amplitude
#define MCCONF_M_ENCODER_COS_AMP 1

// Sine Offset
#define MCCONF_M_ENCODER_SIN_OFFSET 1.65

// Cosine Offset
#define MCCONF_M_ENCODER_COS_OFFSET 1.65

// Sin/Cos Filter Constant
#define MCCONF_M_ENCODER_SINCOS_FILTER 0.5

// Sensor Port Mode
#define MCCONF_M_SENSOR_PORT_MODE 7

// Invert Motor Direction
#define MCCONF_M_INVERT_DIRECTION 0

// DRV8301 OC Mode
#define MCCONF_M_DRV8301_OC_MODE 0

// DRV8301 OC Adjustment
#define MCCONF_M_DRV8301_OC_ADJ 16

// Minimum Switching Frequency
#define MCCONF_M_BLDC_F_SW_MIN 3000

// Maximum Switching Frequency
#define MCCONF_M_BLDC_F_SW_MAX 35000

// Switching Frequency
#define MCCONF_M_DC_F_SW 25000

// Beta Value for Motor Thermistor
#define MCCONF_M_NTC_MOTOR_BETA 3380

// Auxiliary Output Mode
#define MCCONF_M_OUT_AUX_MODE 0

// Motor Temperature Sensor Type
#define MCCONF_M_MOTOR_TEMP_SENS_TYPE 1

// Coefficient for PTC Motor Thermistor
#define MCCONF_M_PTC_MOTOR_COEFF 0.25

// Custom NTC/PTC Resistance
#define MCCONF_M_NTCX_PTCX_RES 10000

// Custom NTC/PTC Base Temperature
#define MCCONF_M_NTCX_PTCX_BASE_TEMP 25

// Hall Sensor Extra Samples
#define MCCONF_M_HALL_EXTRA_SAMPLES 1

// Battery Filter Constant
#define MCCONF_M_BATT_FILTER_CONST 70

// Motor Poles
#define MCCONF_SI_MOTOR_POLES 8

// Gear Ratio
#define MCCONF_SI_GEAR_RATIO 7

// Wheel Diameter
#define MCCONF_SI_WHEEL_DIAMETER 0.714

// Battery Type
#define MCCONF_SI_BATTERY_TYPE 0

// Battery Cells Series
#define MCCONF_SI_BATTERY_CELLS 13

// Battery Capacity
#define MCCONF_SI_BATTERY_AH 14.5

// Motor No Load Current
#define MCCONF_SI_MOTOR_NL_CURRENT 1

// BMS Type
#define MCCONF_BMS_TYPE 1

// BMS Limit Mode
#define MCCONF_BMS_LIMIT_MODE 3

// Temperature Limit Start
#define MCCONF_BMS_T_LIMIT_START 45

// Temperature Limit End
#define MCCONF_BMS_T_LIMIT_END 65

// SOC Limit Start
#define MCCONF_BMS_SOC_LIMIT_START 0.05

// SOC Limit End
#define MCCONF_BMS_SOC_LIMIT_END 0

// Forward CAN to Local
#define MCCONF_BMS_FWD_CAN_MODE 0

// MCCONF_LUNA_M600_H_
#endif

